build-aux/snap/snapcraft.yaml

```diff
@@ -28,172 +29,272 @@ license: GPL-3.0
+           # and libraries from snapd.
+           lib/snapd/snap-exec|lib/snapd/snapctl)
+             export CGO_ENABLED=0
+             GO_LD_FLAGS=()
+             ;;
+           lib/snapd/snap-update-ns)
+             export CGO_ENABLED=1
+             GO_LD_FLAGS=(--ldflags '-extldflags "-static"')
+             ;;
+           *)
+             export CGO_ENABLED=1
+             GO_LD_FLAGS=()
+             ;;
+        esac
+
+        case "${cmd}" in
+          bin/snap)
+            TAGS=(nomanagers)
+            case "$(craftctl get version)" in
+              1337.*)
+                TAGS+=(withtestkeys faultinject)
+                ;;
+            esac
+            ;;
+          *)
+            TAGS=()
+            case "$(craftctl get version)" in
+              1337.*)
+                TAGS+=(withtestkeys withbootassetstesting faultinject)
+                ;;
+            esac
+            ;;
+        esac
+
+        go build -mod=vendor -tags "${TAGS[*]}" "${GO_LD_FLAGS[@]}" ${EXTRA_GO_FLAGS-} -o "${CRAFT_PART_INSTALL}/usr/${cmd}" "github.com/snapcore/snapd/cmd/$(basename ${cmd})"
+        unset CGO_ENABLED
+      done
+
+      make -C data -j"${CRAFT_PARALLEL_BUILD_COUNT}" all
+      # wrappers/core18.go expects units in fixed paths so even when
+      # systemd package will those, they have to continue to match our
+      # paths. We force the value here in case we change the default.
+      make -C data install DESTDIR="${CRAFT_PART_INSTALL}" SYSTEMDUSERUNITDIR=/usr/lib/systemd/user SYSTEMDSYSTEMUNITDIR=/lib/systemd/system
+      # UC depends on this name (symlink)
+      mv "${CRAFT_PART_INSTALL}/etc/profile.d/snapd.sh" "${CRAFT_PART_INSTALL}/etc/profile.d/apps-bin-path.sh"
+
+      install -Dm644 data/info "${CRAFT_PART_INSTALL}/usr/lib/snapd/info"
+      install -Dm644 data/preseed.json "${CRAFT_PART_INSTALL}/usr/lib/snapd/preseed.json"
+
+      install -Dm644 -t "${CRAFT_PART_INSTALL}/usr/lib/snapd" data/completion/bash/complete.sh
+      install -Dm644 -t "${CRAFT_PART_INSTALL}/usr/share/bash-completion/completions" ./data/completion/bash/snap
+
+      if sh -x ./mkversion.sh --output-only | grep "dirty"; then
+        mkdir -p "${CRAFT_PART_INSTALL}/usr/lib/snapd"
+        (
+          echo "dirty git tree during build detected:"
+          git status
+          git diff
+        ) > "${CRAFT_PART_INSTALL}/usr/lib/snapd/dirty-git-tree-info.txt"
+      fi
+
+  fix-linker:
+    plugin: nil
+    build-packages:
+      - patchelf
+      - python3-pyelftools
+    after:
+      # Must run after everything else
+      - snapd
+      - runtime
+      - apparmor
+      - dynamic-linker
+    <<:
+      - *dynamic-linker
+    override-prime: |
+      find "${CRAFT_PRIME}" -type f -exec python3 "${CRAFT_PROJECT_DIR}/build-aux/snap/local/is-exec.py" {} ";" -print -exec patchelf --set-interpreter "/snap/snapd/current/usr/lib/${CRAFT_ARCH_TRIPLET_BUILD_FOR}/${DYNAMIC_LINKER}" {} ";"
```

Comment:
Could we roll this into individual steps (better dependency/reuse) and keep this as a a sanity check that fails if something is linked wrong, perhaps?

Reply:
I have added a commit that refactors it. Tell me what you think.